ts/src/pro/toobit.ts

```diff
@@ -0,0 +1,1183 @@
+        // ]
+        //
+        const subscriptions = Object.keys (client.subscriptions);
+        const accountType = subscriptions[0];
+        if (this.positions === undefined) {
+            this.positions = {};
+        }
+        if (!(accountType in this.positions)) {
+            this.positions[accountType] = new ArrayCacheBySymbolBySide ();
+        }
+        const cache = this.positions[accountType];
+        const newPositions = [];
+        for (let i = 0; i < message.length; i++) {
+            const rawPosition = message[i];
+            const position = this.parseWsPosition (rawPosition);
+            const timestamp = this.safeInteger (rawPosition, 'E');
+            position['timestamp'] = timestamp;
+            position['datetime'] = this.iso8601 (timestamp);
+            newPositions.push (position);
+            cache.append (position);
+        }
+        const messageHashes = this.findMessageHashes (client, accountType + ':positions::');
+        for (let i = 0; i < messageHashes.length; i++) {
+            const messageHash = messageHashes[i];
+            const parts = messageHash.split ('::');
+            const symbolsString = parts[1];
+            const symbols = symbolsString.split (',');
+            const positions = this.filterByArray (newPositions, 'symbol', symbols, false);
+            if (!this.isEmpty (positions)) {
+                client.resolve (positions, messageHash);
+            }
+        }
+        client.resolve (newPositions, accountType + ':positions');
+    }
+
+    parseWsPosition (position, market = undefined) {
+        const marketId = this.safeString (position, 's');
+        return this.safePosition ({
+            'info': position,
+            'id': undefined,
+            'symbol': this.safeSymbol (marketId, undefined),
+            'notional': this.omitZero (this.safeString (position, 'pv')),
+            'marginMode': this.safeStringLower (position, 'mt'),
+            'liquidationPrice': this.safeString (position, 'f'),
+            'entryPrice': this.safeString (position, 'p'),
+            'unrealizedPnl': this.safeString (position, 'up'),
+            'realizedPnl': this.safeNumber (position, 'r'),
+            'percentage': undefined,
+            'contracts': undefined,
+            'contractSize': undefined,
+            'markPrice': undefined,
+            'side': this.safeStringLower (position, 'S'),
+            'hedged': undefined,
+            'timestamp': undefined,
+            'datetime': undefined,
+            'maintenanceMargin': undefined,
+            'maintenanceMarginPercentage': undefined,
+            'collateral': undefined,
+            'initialMargin': this.omitZero (this.safeString (position, 'm')),
+            'initialMarginPercentage': undefined,
+            'leverage': this.safeString (position, 'v'),
+            'marginRatio': undefined,
+        });
+    }
+
+    async authenticate (params = {}) {
+        const time = this.milliseconds ();
```

Comment:
@ttodua here multiple private watchX calls will trigger N requests which is not ideal